core/src/worker/workflow/workflow_stream.rs

```diff
@@ -378,6 +377,20 @@ impl WFStream {
             if let Some(rh) = self.runs.get_mut(run_id) {
                 // Attempt to produce the next activation if needed
                 res = rh.check_more_activations();
+                // If there's no more work and we reported workflow completion to server, evict.
+                if res.is_none()
+                    && rh.workflow_is_finished()
+                    && matches!(report.wft_report_status, WFTReportStatus::Reported { .. })
+                {
+                    res = rh
+                        .request_eviction(RequestEvictMsg {
+                            run_id: run_id.to_string(),
+                            message: "Workflow completed".to_string(),
+                            reason: EvictionReason::WorkflowExecutionEnding,
+                            auto_reply_fail_tt: None,
+                        })
+                        .into_run_update_resp()
+                }
```

Comment:
This is the actual implementation. Everything else is pretty much just test fixing or cleanup / refactor.